pkg/registries/registries.go

```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
 	return match
 }
 
-// rdmContainsARealMirror returns true if set.Mirrors contains at least one entry that is not set.Source.
-func rdmContainsARealMirror(set *apioperatorsv1alpha1.RepositoryDigestMirrors) bool {
-	for _, mirror := range set.Mirrors {
-		if mirror != set.Source {
+// mirrorsContainsARealMirror returns true if mirrors contains at least one entry that is not source.
+func mirrorsContainsARealMirror(source string, mirrors []apicfgv1.ImageMirror) bool {
+	for _, mirror := range mirrors {
+		if string(mirror) != source {
 			return true
 		}
 	}
 	return false
 }
 
-// mergedMirrorSets processes icspRules and returns a set of RepositoryDigestMirrors, one for each Source value,
-// ordered consistently with the preference order of the individual entries (if possible)
-// E.g. given mirror sets (B, C) and (A, B), it will combine them into a single (A, B, C) set.
-func mergedMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) ([]apioperatorsv1alpha1.RepositoryDigestMirrors, error) {
-	disjointSets := map[string]*[]*apioperatorsv1alpha1.RepositoryDigestMirrors{} // Key == Source
-	for _, icsp := range icspRules {
-		for i := range icsp.Spec.RepositoryDigestMirrors {
-			set := &icsp.Spec.RepositoryDigestMirrors[i]
-			if !rdmContainsARealMirror(set) {
-				continue // No mirrors (or mirrors that only repeat the authoritative source) is not really a mirror set.
-			}
-			ds, ok := disjointSets[set.Source]
-			if !ok {
-				ds = &[]*apioperatorsv1alpha1.RepositoryDigestMirrors{}
-				disjointSets[set.Source] = ds
-			}
-			*ds = append(*ds, set)
-		}
+// mirrorSet collects data from mirror setting CRDs (ImageDigestMirrorSet, ImageTagMirrorSet)
+type mirrorSets struct {
+	disjointSets      map[string]*[][]string // Key == Source
+	mirrorBlockSource map[string]bool        // key == Source
+}
+
+func newMirrorSets() *mirrorSets {
+	return &mirrorSets{
+		disjointSets:      map[string]*[][]string{},
+		mirrorBlockSource: map[string]bool{},
 	}
+}
 
-	// Sort the sets of mirrors by Source to ensure deterministic output
+// sourceList collects the mirror sources and returns them sorted in increasing order
+func (sets *mirrorSets) sourceList() []string {
 	sources := []string{}
-	for key := range disjointSets {
+	for key := range sets.disjointSets {
 		sources = append(sources, key)
 	}
 	sort.Strings(sources)
+	return sources
+}
+
+func (sets *mirrorSets) addMirrorSets(source string, mirrorSourcePolicy apicfgv1.MirrorSourcePolicy, mirrors []apicfgv1.ImageMirror) {
```

Comment:
```suggestion
func (sets *mirrorSets) addMirrorSet(source string, mirrorSourcePolicy apicfgv1.MirrorSourcePolicy, mirrors []apicfgv1.ImageMirror) {
```
`sets` is multiple sets, this is adding a single one.